File: kevinotron_data/hmc.py
"""
hmc.py — Full HMC with dynamical Wilson fermions on GPU.

Architecture:
  1. Generate gauge momenta (anti-Hermitian traceless)
  2. Generate pseudofermion φ = D η
  3. Leapfrog: alternate momentum + link updates with gauge + fermion forces
  4. Accept/reject via exp(-ΔH)
"""
import os
os.environ['JAX_ENABLE_X64'] = '1'

import jax
import jax.numpy as jnp
import numpy as np
from time import time

from .wilson_dirac import apply_wilson_dirac, GAMMA5
from .cg_solver import cg_solve
from .pseudofermion import generate_pseudofermion, fermion_action, fermion_force


class HMC:
    def __init__(self, Nc=3, Ls=4, Lt=None, beta=6.0, mass=0.1, n_flavors=2,
                 n_steps=20, dt=0.01):
        self.Nc = Nc
        self.Ls = Ls
        self.Lt = Lt or 2 * Ls
        self.beta = beta
        self.mass = mass
        self.n_flavors = n_flavors
        self.n_steps = n_steps
        self.dt = dt
        
        print(f"HMC: SU({Nc}) Ls={Ls} Lt={self.Lt} β={beta} m={mass} Nf={n_flavors}")
        print(f"     dt={dt} n_steps={n_steps} traj_length={dt*n_steps:.2f}")
    
    def cold_start(self):
        """Initialize identity gauge field."""
        U = jnp.broadcast_to(
            jnp.eye(self.Nc, dtype=jnp.complex128),
            (4, self.Ls, self.Ls, self.Ls, self.Lt, self.Nc, self.Nc)
        ).copy()
        return U
    
    def plaquette(self, U):
        """Average plaquette."""
        p = 0.0
        count = 0
        for mu in range(4):
            for nu in range(mu+1, 4):
                U_mu = U[mu]
                U_nu = U[nu]
                U_mu_shifted = jnp.roll(U_mu, -1, axis=nu)  # U_μ(x+ν)  -- wait, wrong
                # Actually need U_ν(x+μ), U_μ(x+ν), etc.
                U_nu_at_xmu = jnp.roll(U_nu, -1, axis=mu)  # U_ν(x+μ)
                U_mu_at_xnu = jnp.roll(U_mu, -1, axis=nu)  # U_μ(x+ν)
                
                plaq = jnp.einsum('...ab,...bc,...dc,...da->...', 
                                  U_mu, U_nu_at_xmu, 
                                  jnp.conj(U_mu_at_xnu).swapaxes(-2,-1),
                                  jnp.conj(U_nu).swapaxes(-2,-1))
                p += jnp.mean(plaq.real) / self.Nc
                count += 1
        return float(p / count)
    
    def gauge_action(self, U):
        """S_gauge = (β/Nc) Σ (Nc - Re Tr P)."""
        p = self.plaquette(U)
        n_plaq = 6 * self.Ls**3 * self.Lt
        return self.beta * n_plaq * (1.0 - p)
    
    def random_momenta(self, key):
        """Generate random anti-Hermitian traceless momenta."""
        shape = (4, self.Ls, self.Ls, self.Ls, self.Lt, self.Nc, self.Nc)
        Nc = self.Nc
        
        # Random complex matrix
        k1, k2 = jax.random.split(key)
        M = jax.random.normal(k1, shape, dtype=jnp.float64) + \
            1j * jax.random.normal(k2, shape, dtype=jnp.float64)
        M = M / jnp.sqrt(2.0)
        
        # Anti-Hermitian: (M - M†)/2
        M = (M - jnp.conj(M.swapaxes(-2, -1))) / 2
        
        # Traceless
        tr = jnp.trace(M, axis1=-2, axis2=-1)[..., None, None] / Nc
        M = M - tr * jnp.eye(Nc, dtype=jnp.complex128)
        
        return M
    
    def kinetic_energy(self, P):
        """T = -Tr(P²) summed over all links."""
        P2 = jnp.einsum('...ab,...bc->...ac', P, P)
        return float(-jnp.sum(jnp.trace(P2, axis1=-2, axis2=-1)).real)
    
    def gauge_force(self, U):
        """F_gauge = -(β/Nc) × Ta(U × staple_sum)."""
        Nc = self.Nc
        beta_N = self.beta / Nc
        F = jnp.zeros_like(U)
        
        for mu in range(4):
            staple = jnp.zeros((self.Ls, self.Ls, self.Ls, self.Lt, Nc, Nc), 
                              dtype=jnp.complex128)
            for nu in range(4):
                if nu == mu:
                    continue
                # Forward staple
                U_nu_at_xmu = jnp.roll(U[nu], -1, axis=mu)
                U_mu_at_xnu = jnp.roll(U[mu], -1, axis=nu)
                s_fwd = jnp.einsum('...ab,...bc,...dc->...ad',
                                   U_nu_at_xmu,
                                   jnp.conj(U_mu_at_xnu).swapaxes(-2,-1),
                                   jnp.conj(U[nu]).swapaxes(-2,-1))
                
                # Backward staple
                U_nu_at_xmu_mnu = jnp.roll(jnp.roll(U[nu], -1, axis=mu), 1, axis=nu)
                U_mu_at_xmnu = jnp.roll(U[mu], 1, axis=nu)
                U_nu_at_xmnu = jnp.roll(U[nu], 1, axis=nu)
                s_bwd = jnp.einsum('...ba,...dc,...cd->...ad',
                                   jnp.conj(U_nu_at_xmu_mnu),
                                   jnp.conj(U_mu_at_xmnu),
                                   U_nu_at_xmnu)  # this needs fixing
                # Actually: U†_ν(x+μ-ν) × U†_μ(x-ν) × U_ν(x-ν)
                s_bwd = jnp.einsum('...ba,...dc,...de->...ae',
                                   jnp.conj(U_nu_at_xmu_mnu),
                                   jnp.conj(U_mu_at_xmnu),
                                   U_nu_at_xmnu)
                
                staple = staple + s_fwd + s_bwd
            
            V = jnp.einsum('...ab,...bc->...ac', U[mu], staple)
            V = V * beta_N
            
            # Anti-Hermitian traceless projection
            V_ah = (V - jnp.conj(V.swapaxes(-2,-1))) / 2
            tr = jnp.trace(V_ah, axis1=-2, axis2=-1)[..., None, None] / Nc
            V_ah = V_ah - tr * jnp.eye(Nc, dtype=jnp.complex128)
            
            F = F.at[mu].set(V_ah)
        
        return F
    
    def update_momenta(self, P, F_gauge, F_ferm, dt):
        """P -= dt × (F_gauge + F_ferm)."""
        return P - dt * (F_gauge + F_ferm)
    
    def update_links(self, U, P, dt):
        """U = expm(dt × P) × U."""
        Nc = self.Nc
        dtP = dt * P
        # Taylor-12 expm
        exp_P = jnp.broadcast_to(jnp.eye(Nc, dtype=jnp.complex128), dtP.shape).copy()
        power = jnp.broadcast_to(jnp.eye(Nc, dtype=jnp.complex128), dtP.shape).copy()
        for k in range(1, 13):
            power = jnp.einsum('...ab,...bc->...ac', power, dtP) / k
            exp_P = exp_P + power
        
        U_new = jnp.einsum('...ab,...bc->...ac', exp_P, U)
        
        # Reunitarize (Gram-Schmidt)
        for j in range(Nc):
            col = U_new[..., :, j]
            for i in range(j):
                prev = U_new[..., :, i]
                proj = jnp.sum(jnp.conj(prev) * col, axis=-1, keepdims=True)
                col = col - proj * prev
            norm = jnp.sqrt(jnp.sum(jnp.abs(col)**2, axis=-1, keepdims=True))
            U_new = U_new.at[..., :, j].set(col / jnp.maximum(norm, 1e-15))
        
        # Fix det = 1 (approximate)
        return U_new
    
    def trajectory(self, U, key):
        """One HMC trajectory with gauge + fermion forces."""
        k1, k2, k3 = jax.random.split(key, 3)
        
        U_old = U.copy()
        
        # Generate momenta
        P = self.random_momenta(k1)
        
        # Generate pseudofermion
        phi, _ = generate_pseudofermion(U, self.mass, self.Ls, self.Lt, k2)
        
        # Initial Hamiltonian
        T_old = self.kinetic_energy(P)
        S_gauge_old = self.gauge_action(U)
        S_ferm_old, _, _ = fermion_action(U, phi, self.mass, self.Ls, self.Lt)
        H_old = T_old + S_gauge_old + S_ferm_old
        
        # Leapfrog
        F_gauge = self.gauge_force(U)
        F_ferm = fermion_force(U, phi, self.mass, self.Ls, self.Lt)
        
        # Half step momenta
        P = self.update_momenta(P, F_gauge, F_ferm, 0.5 * self.dt)
        
        for step in range(self.n_steps - 1):
            U = self.update_links(U, P, self.dt)
            F_gauge = self.gauge_force(U)
            F_ferm = fermion_force(U, phi, self.mass, self.Ls, self.Lt)
            P = self.update_momenta(P, F_gauge, F_ferm, self.dt)
        
        U = self.update_links(U, P, self.dt)
        F_gauge = self.gauge_force(U)
        F_ferm = fermion_force(U, phi, self.mass, self.Ls, self.Lt)
        P = self.update_momenta(P, F_gauge, F_ferm, 0.5 * self.dt)
        
        # Final Hamiltonian
        T_new = self.kinetic_energy(P)
        S_gauge_new = self.gauge_action(U)
        S_ferm_new, _, _ = fermion_action(U, phi, self.mass, self.Ls, self.Lt)
        H_new = T_new + S_gauge_new + S_ferm_new
        
        dH = H_new - H_old
        
        # Accept/reject
        r = jax.random.uniform(k3, dtype=jnp.float64)
        accepted = float(r) < min(1.0, np.exp(-dH))
        
        if not accepted:
            U = U_old
        
        return U, accepted, dH
    
    def generate(self, n_traj=100, n_therm=50):
        """Run HMC and return thermalized configurations."""
        U = self.cold_start()
        key = jax.random.PRNGKey(42)
        
        n_accept = 0
        for traj in range(n_traj):
            key, subkey = jax.random.split(key)
            t0 = time()
            U, accepted, dH = self.trajectory(U, subkey)
            dt_wall = time() - t0
            
            if accepted:
                n_accept += 1
            
            if (traj + 1) % 5 == 0:
                p = self.plaquette(U)
                print(f"  traj {traj+1:>4}: dH={dH:>+8.4f} "
                      f"acc={n_accept}/{traj+1} ({100*n_accept/(traj+1):.0f}%) "
                      f"P={p:.6f} [{dt_wall:.1f}s]")
        
        print(f"HMC done: {n_accept}/{n_traj} accepted ({100*n_accept/n_traj:.0f}%)")
        return U


if __name__ == '__main__':
    hmc = HMC(Nc=3, Ls=4, beta=5.6, mass=0.5, n_steps=10, dt=0.01)
    U = hmc.generate(n_traj=20, n_therm=0)

File: kevinotron_data/hmc.rs
// hmc.rs — Hybrid Monte Carlo (CORRECTED)
//
// KEY MATH:
// Momenta π ∈ Lie algebra (anti-Hermitian traceless for SU(N), antisymmetric for SO/G₂/F₄)
// Kinetic energy: T = -Tr(π²) (positive since π anti-Hermitian)
// Gauge force: F = project_to_algebra(-(β/d) × U × staple_sum)
//   For SU(N): F = (V - V†)/2 - Tr(V-V†)/(2N) I where V = (β/N) U K
//   For real:  F = (V - V^T)/2 where V = (β/d) U K
// Update: U_new = exp(dt × π) × U_old (π is already in the algebra)

use crate::groups::{GaugeGroup, LinkData, RngWrapper};
use crate::lattice::Lattice4D;
use rand::Rng;
use rand::RngCore;
use rand_distr::StandardNormal;

pub struct HmcConfig {
    pub n_steps: usize,
    pub dt: f64,
    pub beta: f64,
}

/// Generate random momentum in the Lie algebra
/// Complex groups: π = anti-Hermitian traceless (2*d*d f64)
/// Real groups: π = antisymmetric (d*d f64)
fn random_algebra_element(group: &dyn GaugeGroup, rng: &mut dyn RngCore) -> LinkData {
    // Use random_near_id with large epsilon, then take log ≈ the algebra element
    // Actually simpler: generate random anti-Hermitian traceless matrix directly
    let d = group.dim_fund();
    let mut w = RngWrapper(rng);

    if group.is_complex() {
        let mut m = vec![0.0f64; 2 * d * d];
        // Fill upper triangle with random complex, reflect to lower
        for i in 0..d {
            for j in (i + 1)..d {
                let re: f64 = w.sample::<f64, _>(StandardNormal);
                let im: f64 = w.sample::<f64, _>(StandardNormal);
                let idx_ij = 2 * (i * d + j);
                let idx_ji = 2 * (j * d + i);
                // Anti-Hermitian: M_ij = a+ib, M_ji = -a+ib (so M† = -M)
                m[idx_ij] = re;        // Re M_ij
                m[idx_ij + 1] = im;    // Im M_ij
                m[idx_ji] = -re;       // Re M_ji = -Re M_ij
                m[idx_ji + 1] = im;    // Im M_ji = Im M_ij (since conj swap + negate)
            }
            // Diagonal: purely imaginary for anti-Hermitian
            let diag_val: f64 = w.sample::<f64, _>(StandardNormal);
            m[2 * (i * d + i) + 1] = diag_val; // Im M_ii
        }
        // Make traceless: subtract Tr(M)/d from diagonal
        let mut tr_im = 0.0;
        for i in 0..d {
            tr_im += m[2 * (i * d + i) + 1];
        }
        tr_im /= d as f64;
        for i in 0..d {
            m[2 * (i * d + i) + 1] -= tr_im;
        }
        m
    } else {
        let mut m = vec![0.0f64; d * d];
        // Antisymmetric: M_ij = -M_ji
        for i in 0..d {
            for j in (i + 1)..d {
                let val: f64 = w.sample::<f64, _>(StandardNormal);
                m[i * d + j] = val;
                m[j * d + i] = -val;
            }
        }
        m
    }
}

/// Kinetic energy: T = -Tr(π²) for anti-Hermitian π
/// For complex: -Tr(π²) = -Σ_{ij} π_ij π_ji = -Σ_ij π_ij (-π_ij*) = Σ |π_ij|²
/// For real: -Tr(π²) = -Σ_ij π_ij π_ji = Σ_ij π_ij² (antisymmetric)
fn kinetic_energy(momenta: &[LinkData], is_complex: bool) -> f64 {
    let mut t = 0.0;
    for p in momenta {
        if is_complex {
            // -Tr(π²) = Σ |π_ij|² = Σ (re² + im²)
            for &x in p.iter() {
                t += x * x;
            }
        } else {
            // -Tr(π²) = 2 Σ_{i<j} π_ij² (each off-diag appears twice with opposite sign)
            for &x in p.iter() {
                t += x * x;
            }
        }
    }
    0.5 * t
}

/// Project V = (β/d) U K onto the Lie algebra
/// Complex: Ta(V) = (V - V†)/2 - Tr(V-V†)/(2d) I
/// Real: Ta(V) = (V - V^T)/2
fn project_force(group: &dyn GaugeGroup, v: &[f64]) -> LinkData {
    let d = group.dim_fund();

    if group.is_complex() {
        let mut f = vec![0.0f64; 2 * d * d];
        // V† stored as conjugate transpose
        let v_dag = crate::groups::cmat_dagger(v, d);
        // (V - V†)/2
        for i in 0..f.len() {
            f[i] = 0.5 * (v[i] - v_dag[i]);
        }
        // Subtract trace
        let mut tr_re = 0.0;
        let mut tr_im = 0.0;
        for i in 0..d {
            let idx = 2 * (i * d + i);
            tr_re += f[idx];
            tr_im += f[idx + 1];
        }
        tr_re /= d as f64;
        tr_im /= d as f64;
        for i in 0..d {
            let idx = 2 * (i * d + i);
            f[idx] -= tr_re;
            f[idx + 1] -= tr_im;
        }
        f
    } else {
        let mut f = vec![0.0f64; d * d];
        // (V - V^T)/2
        for i in 0..d {
            for j in 0..d {
                f[i * d + j] = 0.5 * (v[i * d + j] - v[j * d + i]);
            }
        }
        f
    }
}

/// Gauge action
fn gauge_action(lat: &Lattice4D, group: &dyn GaugeGroup, beta: f64) -> f64 {
    let p = lat.plaquette(group);
    let d = group.dim_fund() as f64;
    let n_plaq = 6 * lat.ls.pow(3) * lat.lt;
    beta * n_plaq as f64 * (1.0 - p)
}


/// Update all momenta: p -= dt × F(U)
fn update_momenta(
    lat: &Lattice4D,
    momenta: &mut [LinkData],
    group: &dyn GaugeGroup,
    beta_d: f64,
    dt: f64,
) {
    let ls = lat.ls; let lt = lat.lt;
    let mut idx = 0;
    for x0 in 0..ls {
        for x1 in 0..ls {
            for x2 in 0..ls {
                for x3 in 0..lt {
                    for mu in 0..4 {
                        let site = [x0,x1,x2,x3];
                        let u = lat.get(site, mu);
                        let k = lat.staple_sum(group, site, mu);
                        let mut v = group.mul(u, &k);
                        for x in v.iter_mut() { *x *= beta_d; }
                        let f = project_force(group, &v);
                        for i in 0..momenta[idx].len() {
                            momenta[idx][i] -= dt * f[i];
                        }
                        idx += 1;
                    }
                }
            }
        }
    }
}

/// Update all links: U = exp(dt × π) × U
fn update_links(
    lat: &mut Lattice4D,
    momenta: &[LinkData],
    group: &dyn GaugeGroup,
    dt: f64,
) {
    let ls = lat.ls; let lt = lat.lt;
    let d = group.dim_fund();
    let is_complex = group.is_complex();
    let mut idx = 0;
    for x0 in 0..ls {
        for x1 in 0..ls {
            for x2 in 0..ls {
                for x3 in 0..lt {
                    for mu in 0..4 {
                        let site = [x0,x1,x2,x3];
                        let u = lat.get(site, mu).to_vec();
                        let p = &momenta[idx];
                        // exp(dt π) via Taylor-12 (matching Metropolis precision)
                        let exp_p;
                        if is_complex {
                            let mut s = p.clone();
                            for x in s.iter_mut() { *x *= dt; }
                            exp_p = crate::groups::cmat_expm_taylor12(&s, d);
                        } else {
                            // Real groups: Taylor-12 manually
                            let mut s = p.clone();
                            for x in s.iter_mut() { *x *= dt; }
                            let mut sum = crate::groups::rmat_identity(d);
                            let mut pow = crate::groups::rmat_identity(d);
                            for k in 1..=12 {
                                pow = crate::groups::rmat_mul(&pow, &s, d);
                                let inv_k = 1.0 / k as f64;
                                for i in 0..pow.len() { pow[i] *= inv_k; }
                                for i in 0..sum.len() { sum[i] += pow[i]; }
                            }
                            exp_p = sum;
                        }
                        let u_new = if is_complex {
                            crate::groups::cmat_mul(&exp_p, &u, d)
                        } else {
                            crate::groups::rmat_mul(&exp_p, &u, d)
                        };
                        lat.set(site, mu, &group.reproject(&u_new));
                        idx += 1;
                    }
                }
            }
        }
    }
}

/// One HMC trajectory
pub fn hmc_trajectory(
    lat: &mut Lattice4D,
    group: &dyn GaugeGroup,
    cfg: &HmcConfig,
    rng: &mut dyn RngCore,
) -> (bool, f64) {
    let ls = lat.ls;
    let lt = lat.lt;
    let n_links = 4 * ls * ls * ls * lt;
    let is_complex = group.is_complex();
    let beta_d = cfg.beta / group.beta_norm();

    // Save old config
    let mut old_links = Vec::with_capacity(n_links);
    for x0 in 0..ls {
        for x1 in 0..ls {
            for x2 in 0..ls {
                for x3 in 0..lt {
                    for mu in 0..4 {
                        old_links.push(lat.get([x0,x1,x2,x3], mu).to_vec());
                    }
                }
            }
        }
    }

    // Generate algebra momenta
    let mut momenta: Vec<LinkData> = (0..n_links)
        .map(|_| random_algebra_element(group, rng))
        .collect();

    // Initial Hamiltonian
    let t_old = kinetic_energy(&momenta, is_complex);
    let s_old = gauge_action(lat, group, cfg.beta);
    let h_old = t_old + s_old;

    // Leapfrog: proper V-V-T-V-V scheme (Verlet)
    // Step 0: half-step momenta
    update_momenta(lat, &mut momenta, group, beta_d, 0.5 * cfg.dt);
    
    // Steps 1..n-1: full link + full momenta
    for _step in 0..cfg.n_steps - 1 {
        update_links(lat, &momenta, group, cfg.dt);
        update_momenta(lat, &mut momenta, group, beta_d, cfg.dt);
    }
    
    // Final: full link + half momenta
    update_links(lat, &momenta, group, cfg.dt);
    update_momenta(lat, &mut momenta, group, beta_d, 0.5 * cfg.dt);

    // Final Hamiltonian
    let t_new = kinetic_energy(&momenta, is_complex);
    let s_new = gauge_action(lat, group, cfg.beta);
    let h_new = t_new + s_new;
    let delta_h = h_new - h_old;

    // WARNING: HMC is only correct for groups where the full anti-Hermitian
    // traceless (complex) or antisymmetric (real) space IS the Lie algebra.
    // This holds for SU(N) and SO(N), but NOT for G₂⊂SO(7), Sp(4)⊂SU(4),
    // F₄⊂SO(26), E₆⊂SU(27). For these groups, use Metropolis mode instead.
    // TODO: implement group-specific algebra projection using stored generators.

    // Accept/reject
    let mut w = RngWrapper(rng);
    let r: f64 = w.gen();
    let accepted = r < (-delta_h).exp().min(1.0);

    if !accepted {
        let mut link_idx = 0;
        for x0 in 0..ls {
            for x1 in 0..ls {
                for x2 in 0..ls {
                    for x3 in 0..lt {
                        for mu in 0..4 {
                            lat.set([x0,x1,x2,x3], mu, &old_links[link_idx]);
                            link_idx += 1;
                        }
                    }
                }
            }
        }
    }

    (accepted, delta_h)
}
